### PR TITLE
Add a WordList protection

### PR DESCRIPTION
**Pull request by Alch-Emi**
Add a protection for banning users based off of the use of certain blocked words.

Customization allows for modifying the word list in the config, using case-insensitive regexs to detect malicious messages.  Additionally, can be configured to only apply to messages set within a certain time period after a user joins, a feature which is beneficial in environments where certain words may come up naturally in conversation, but are almost always a red flag if used immediately on joining.  This does not add users to a ban list.

A feature like this stands to benefit many room admins.  In my experience, much of room administration is just noticing trolls and banning that quickly.  Many trolls make this blatantly easy by using inflammatory language immediately on joining, to the point that automating it is trivial and likely to have a high accuracy rate.

I'd love to hear your thoughts

```
Signed-off-by: Emi Tatsuo <emi@alchemi.dev>
```

### CODE REVIEW

File: config/default.yaml [hunk 1]
```diff
@@ -107,6 +107,24 @@ commands:
...
+  # blocked words shortly after joining.
+  wordlist:
+    # A list of words which should be monitored by the bot.  These will match if any part
+    # of the word is present in the message in any case.  e.g. "poop" also matches
+    # "poOPYHEad".  Additionally, regular expressions can be used.
+    words:
+      - "nigger"
+      - "faggot"
+      - "tranny"
+      - "retard"
```
Comment (turt2live) [inline config/default.yaml:122]: For better or worse, we try to keep this project neutral to any particular kinds of abuse. We also definitely do not try and expose admins of the bot to abuse themselves, even if it is an example.

```suggestion
    # A list of words which should be monitored by the bot. These will match if any part
    # of the word is present in the message in any case. Regular expressions are also supported.
    words:
      - "CaSE"
      - "inSEnsITIVe"
```

File: config/default.yaml [hunk 2]
```diff
@@ -107,6 +107,24 @@ commands:
...
+    # How long after a user joins the server should the bot monitor their messages.  After
+    # this time, users can say words from the wordlist without being banned automatically.
+    # Set to zero to disable (users will always be banned if they say a bad word)
+    minutesBeforeTrusting: 20
```
Comment (turt2live) [inline config/default.yaml:126]:
```suggestion
    
    # How long after a user joins the room should the bot monitor their messages.  After
    # this time, users can say words from the wordlist without being banned automatically.
    # Set to zero to disable (users will always be banned if they say a bad word)
    minutesBeforeTrusting: 20
```

File: src/config.ts [hunk 3]
```diff
@@ -89,6 +95,12 @@ const defaultConfig: IConfig = {
         allowNoPrefix: false,
         additionalPrefixes: [],
     },
+    protections: {
+        wordlist: {
+            words: ["nigger", "faggot", "tranny", "retard"],
```
Comment (turt2live) [inline src/config.ts:100]: the default should be empty in this case:
```suggestion
            words: [],
```

File: src/protections/WordList.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2019, 2020 The Matrix.org Foundation C.I.C.
```
Comment (turt2live) [inline src/protections/WordList.ts:2]: Unless you're actually assigning copyright to the foundation, I suspect this should be your name. For example:

`Copyright 2020 Emi Tatsuo`

File: src/protections/WordList.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    public get name(): string {
+        return 'WordList';
+    }
+
+    public async handleEvent(mjolnir: Mjolnir, roomId: string, event: any): Promise<any> {
+
+        const content = event['content'] || {};
+        const mbt = config.protections.wordlist.minutesBeforeTrusting;
```
Comment (turt2live) [inline src/protections/WordList.ts:44]: we have the luxury of large line limits: naming this something like `minBeforeTrusted` is entirely reasonable.

**Comment by turt2live** [comment]
Overall this is looking like a great addition - thank you!

**Comment by Alch-Emi** [comment]
Ha remember when I wrote this and then totally forgot about it.  Woops.  Anyway, I applied those changes now.  lmk what you think

File: config/default.yaml [hunk 6]
```diff
@@ -107,6 +107,25 @@ commands:
...
     - "mjolnir_bot"
 
+# Configuration specific to certain toggleable protections
+protections:
+  # Configuration for the wordlist plugin, which can ban users based if they say certain
+  # blocked words shortly after joining.
+  wordlist:
+    # A list of words which should be monitored by the bot.  These will match if any part
+    # of the word is present in the message in any case.  e.g. "poop" also matches
+    # "poOPYHEad".  Additionally, regular expressions can be used.
```
Comment (turt2live) [inline config/default.yaml:117]:
```suggestion
    # of the word is present in the message in any case.  e.g. "hello" also matches
    # "HEllO".  Additionally, regular expressions can be used.
```

**Review by turt2live — APPROVED** [review]
Thanks for working through this!

**Pull request merged**